sdk/program/src/message/compiled_keys.rs

```diff
@@ -0,0 +1,222 @@
+            .collect();
+
+        CompiledKeys {
+            writable_signer_keys,
+            readonly_signer_keys,
+            writable_non_signer_keys,
+            readonly_non_signer_keys,
+        }
+    }
+
+    pub(crate) fn try_create_message_header(&self) -> Option<MessageHeader> {
+        Some(MessageHeader {
+            num_required_signatures: u8::try_from(
+                self.writable_signer_keys
+                    .len()
+                    .checked_add(self.readonly_signer_keys.len())?,
+            )
+            .ok()?,
+            num_readonly_signed_accounts: u8::try_from(self.readonly_signer_keys.len()).ok()?,
+            num_readonly_unsigned_accounts: u8::try_from(self.readonly_non_signer_keys.len())
+                .ok()?,
+        })
+    }
+
+    pub(crate) fn into_message_static_keys(self) -> Vec<Pubkey> {
+        std::iter::empty()
+            .chain(self.writable_signer_keys)
+            .chain(self.readonly_signer_keys)
+            .chain(self.writable_non_signer_keys)
+            .chain(self.readonly_non_signer_keys)
+            .collect()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use {super::*, crate::instruction::AccountMeta};
+
+    #[test]
+    fn test_compile_with_dup_keys() {
+        let program_id = Pubkey::new_unique();
+        let id0 = Pubkey::new_unique();
+        let id1 = Pubkey::new_unique();
+        let id2 = Pubkey::new_unique();
+        let instruction = Instruction::new_with_bincode(
+            program_id,
+            &0,
+            vec![
+                AccountMeta::new(id0, true),
+                AccountMeta::new_readonly(id1, true),
+                AccountMeta::new(id2, false),
+            ],
+        );
+
+        let keys = CompiledKeys::compile(&[instruction.clone(), instruction], None);
+        assert_eq!(
+            keys,
+            CompiledKeys {
+                writable_signer_keys: vec![id0],
+                readonly_signer_keys: vec![id1],
+                writable_non_signer_keys: vec![id2],
+                readonly_non_signer_keys: vec![program_id],
+            }
+        );
+    }
+
+    #[test]
+    fn test_compile_with_dup_payer() {
+        let program_id = Pubkey::new_unique();
+        let payer = Pubkey::new_unique();
+        let keys = CompiledKeys::compile(
+            &[Instruction::new_with_bincode(
+                program_id,
+                &0,
+                vec![AccountMeta::new_readonly(payer, false)],
+            )],
+            Some(payer),
+        );
+        assert_eq!(
+            keys,
+            CompiledKeys {
+                writable_signer_keys: vec![payer],
+                readonly_non_signer_keys: vec![program_id],
+                ..CompiledKeys::default()
+            }
+        );
+    }
+
+    #[test]
+    fn test_compile_with_dup_signer_mismatch() {
+        let program_id = Pubkey::new_unique();
+        let id0 = Pubkey::new_unique();
+        let keys = CompiledKeys::compile(
```

Comment:
Again, I think this would be more clear with a single instruction with the same account referenced as signing and not

edit: throughout.  rational being that there's no logic in the outer loop, so the extra instructions only muddy the test

Reply:
Agreed, updated.